soroban-env-host/src/crypto/bls12_381.rs

```diff
@@ -0,0 +1,665 @@
+        if points.len() != scalars.len() || points.len() == 0 {
+            return Err(
+                Error::from_type_and_code(ScErrorType::Crypto, ScErrorCode::InvalidInput).into(),
+            );
+        }
+        // The actual logic happens inside msm_bigint_wnaf (ark_ec/variable_base/mod.rs)
+        // under branch negation is cheap.
+        // the unchecked version just skips the length equal check
+        self.charge_budget(ContractCostType::Bls12381G2Msm, Some(points.len() as u64))?;
+        Ok(G2Projective::msm_unchecked(points, scalars))
+    }
+
+    pub(crate) fn map_fp2_to_g2_internal(&self, fp: Fq2) -> Result<G2Affine, HostError> {
+        self.charge_budget(ContractCostType::Bls12381MapFp2ToG2, None)?;
+        let mapper = WBMap::<g2::Config>::new().map_err(|e| {
+            self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InternalError,
+                format!("hash-to-curve error {e}").as_str(),
+                &[],
+            )
+        })?;
+        mapper.map_to_curve(fp).map_err(|e| {
+            self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InternalError,
+                format!("hash-to-curve error {e}").as_str(),
+                &[],
+            )
+        })
+    }
+
+    pub(crate) fn hash_to_g2_internal(
+        &self,
+        domain: &[u8],
+        msg: &[u8],
+    ) -> Result<G2Affine, HostError> {
+        self.charge_budget(ContractCostType::Bls12381HashToG2, Some(msg.len() as u64))?;
+        let mapper = MapToCurveBasedHasher::<
+            Projective<g2::Config>,
+            DefaultFieldHasher<Sha256, 128>,
+            WBMap<g2::Config>,
+        >::new(domain)
+        .map_err(|e| {
+            self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InternalError,
+                format!("hash-to-curve error {e}").as_str(),
+                &[],
+            )
+        })?;
+        mapper.hash(msg.as_ref()).map_err(|e| {
+            self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InternalError,
+                format!("hash-to-curve error {e}").as_str(),
+                &[],
+            )
+        })
+    }
+
+    pub(crate) fn pairing_internal(
+        &self,
+        vp1: &Vec<G1Affine>,
+        vp2: &Vec<G2Affine>,
+    ) -> Result<PairingOutput<Bls12_381>, HostError> {
+        // this check should've been done outside, here is just extra caution
+        if vp1.len() != vp2.len() || vp1.len() == 0 {
+            return Err(
+                Error::from_type_and_code(ScErrorType::Crypto, ScErrorCode::InvalidInput).into(),
+            );
+        }
+        self.charge_budget(ContractCostType::Bls12381Pairing, Some(vp1.len() as u64))?;
+        // This is doing exact same as `Bls12_381::pairing(p, q)`, but avoids
+        // the `unwrap`
+        let mlo = Bls12_381::multi_miller_loop(vp1, vp2);
```

Comment:
How about wrapping the arkworks library call with `panic::catch_unwind` like we do in core?

Reply:
We should not add `catch_unwind` here (or anywhere other than the one already in core) because:
- unwinding is very expensive, in terms of code generated to handle the unwinding as well as the unwinding logic itself (rust heavily optimizes for the panic-less path).
- some panic can leave the host in an unstable state, so it only makes sense to do it if we tear down the host immediately afterwards (which is where it is in core). Between a function failure (here) and returning to core there are still a bunch of logic such as returning from frame, metering, generating diagnostics etc, which could leave the host in a bad state.